### DISCUSSION TITLE
Draft Documents

### DISCUSSION THREAD

**Discussion by jamespohalloran** [tinacms/tinacms#2962] [Experimental]
## The problem

Editors are looking for a way to save "draft" changes, without having those changes immediately published to the production site. 
Since Tina is a "git-based" CMS's that supports branching, this is technically possible right now by creating new branches for each change, however, we have found that the concept of working in branches isn't always intuitive for non-developers. 

## Proposed Solution

We'd like to allow editors to control the draft-state on the document-level. 

To accomplish this, we're proposing implementing a 2-branch system: (`<baseBranch>` & `<baseBranch>-staging` ), so that drafts of documents can be persisted, before being published.

<img width="1062" alt="162043677-7e0fb945-4a17-4bc7-9604-8d266e953046" src="https://user-images.githubusercontent.com/3323181/172690016-815dd920-e9d1-4381-907f-18d40a8637dd.png">

In the editor UI, a document can have one of 3 states:

_Assuming your base branch configured with Tina is "main"._

**Draft**: The document has been created, but not yet published to `main`.

**Published**: The document has been published to `main`, and there are no pending changes.

**Changed**: The document lives on the production `main` branch, however, there are new changes in `main-staging` that haven't been published yet. 

<img width="494" alt="162052732-891f5c66-6ca3-4749-a3dd-e6d06f47db89" src="https://user-images.githubusercontent.com/3323181/172689924-0c5a135a-8c98-4dd1-bb3d-2919bb80b6e5.png">

Assuming your environment is configured to use the `main` branch with Tina; when you enter edit mode, the frontend logic switches the request to now use `main-staging`. Any edits made on this site will all persist to the `main-staging` branch. So when you're in edit mode, you'll see other's pending changes along with your own. To "publish" something, we cherry-pick that document into the `main` branch.

### Keeping the staging branch in-sync

One of the interesting aspects of this is what we'll do when we get pushes on a branch from Github that conflict with the staging branch. We've talked about how we'll need a UI that helps us resolve these, but I think - at least for our first iteration, we can basically merge `main` into `main-staging`, and always let the "current" changes that `main-staging` has _win_ that conflict.

### More UI Mockups:

@spbyrne has done up some early Figma mockups of the editor workflow:

Clickable Figma prototype:
https://www.figma.com/proto/pZIV0YELAhgzpNTC6QXmaB/Tina-Editorial-Workflow-2022?page-id=0%3A1&node-id=34%3A379&viewport=51148%2C15065%2C10.23&scaling=min-zoom&starting-point-node-id=2%3A13

Only certain things are clickable, so here are some quick videos to demonstrate it.

Here's a demo of editing, saving, publishing & viewing revision history:
https://user-images.githubusercontent.com/5075484/162824308-46166eb7-7147-496f-86a3-bc7b2a2b0a7b.mp4

Here's a demo of publishing from the list page, or creating and publishing a new document:
https://user-images.githubusercontent.com/5075484/162824311-0b050774-e23f-4488-bbc9-e0b22cbdede3.mp4

---

At the time of writing, no development work has been started on this project, so we'd love to hear your feedback, good or bad on this approach!

**Comment by korompaiistvan** [comment]
I am setting up a team blog with Tina and found this document while looking for guides/inspiration on how to set up an editorial workflow. This long comment is partly thinking "out loud" but might offer you some insight in what I find important.

### Context
We'll have about a dozen potential Editors on this blog, each working in their own speed on new content. While most of these Editors are technically minded, most of them are not comfortable with git and web technologies. A select group of Editors function as Reviewers (4 or 5), and a very limited number serve as Developers (2 at max), who can edit the HTML-JS-CSS locally and work confidently with git.

There might be a handful of different posts being worked on at a time, each over multiple different commits, over long periods of time.

Once a piece of content is published, it's rarely revisited (apart from "hotfixes", such as typos, corrections, etc.). This is important because our solution will optimize much more heavily for new content rather than revisions.

We will use Netlify to deploy the site (initally SSR, later pivoting to SSG).

### Potential solutions I've thought of
#### Building up release branches manually
1. There are two branches of the repository, `main` and `draft`. The `main` branch is deployed to the production site, while the `draft` is deployed separately.
1. Editors work on their content through the admin page of the draft deployment.
2. Once an Editor deems a certain piece finished, they notify a Developer
3. The notified Developer cherry picks all the relevant commits into a `publish-xyz` branch
4. The publish branch gets a deploy preview in Netlify
5. The Reviewers use the deploy preview to review the changes
6. Any further changes to the content come to the publish branch directly
7. Once fully reviewed the Developer merges the publish branch into main

Pros:
- requires no additional Tina setup (apart from making sure that the two URLs point at the right branches)

Cons:
- heavily relies on the people with the scarcest skills (i.e. Developers)
- cherry picking all the right commits into a publish branch for the different posts is fairly painful (since the commit messages emitted from contextual edits are not very informative).

#### Using a `public` attribute in the schema
1. This setup uses only a `main` branch, the CMS commits directly here
2. The posts have a "Published?" boolean field that can be set by the Editors. Posts with this field set to false, simply do not show up anywhere on the home page (and we prevent them being crawled by indexers)
3. Once the responsible Editor deems a piece of content ready for publishing they notify the Reviewers directly
4. After the review and corrections, the flag is set to True.

Con:
- Relies completely on Editor discipline to set the flag properly
- Each small edit triggers a republish of the production site

Pros:
- Very easy to understand for Editors
- Similar to how Wordpress deals with this problem -> Familiar

#### Copying NetlifyCMS
The only other git-based CMS I have any knowledge of is NetlifyCMS which uses a different model to what you outline
https://preview-auth-doc--netlify-cms-www.netlify.app/docs/editorial-workflow/.

They have a branch per piece of content, so the process for us, using Tina would look like:
1. Editor creates a new branch or notifies the Developers they would like a new branch created for them (let's call it `draft-xyz`)
2. The new branch is picked up by Netlify and gets deployed on its own URL.
3. Tina is used by the Editors on this new URL and it commits back to the `draft-xyz` branch.
4. Once the responsible Editor deems the content ready for publishing they notify the Reviewers directly, sending the URL
5. After the review and corrections, a developer merges the `draft-xyz` branch into `main`

Cons:
- Potentially lot of different branches and corresponding URLs which might be difficult to keep track of.
- Draft URLs are not known upfront so this requires a Tina Cloud configuration change for every new branch if we rely on Netlify's own URL generation

Pros:
- Less manual work for Developers, and what is still needed is lower complexity compared to the manual release branches option.
- Changes on different pieces of content do not mix together -> no need for cherry picking

### Conclusion
For me, right now, the NetlifyCMS route seems to be the easiest to understand for Editors, and least ongoing effort for me, as one of the Developers.

Your solution, outlined in the original post seems to be a better version of the manual release branch approach, where the need for cherry picking the right commits is removed by some automation. That automation would essentially remove both of the drawbacks I have listed, making it an ideal solution.  
he only thing missing for me is the distinction between people who can modify draft content, vs those who can also publish it. To be a complete solution, that should be built in.

**Reply by scottgallant** [reply]
@korompaiistvan , thanks for this write up!

> The only thing missing for me is the distinction between people who can modify draft content, vs those who can also publish it. To be a complete solution, that should be built in.

Can you describe how this would work on your team? What type of people publish but don't modify content on your team?  Does all content need to go through that workflow?   What would and ideal solution look like?

**Reply by korompaiistvan** [reply]
The short answer is we don't even know yet what our ideal editorial flow looks like, much less how much of it is possible with Tina.

The longer answer: all our CMS users are going to be Editors (can edit content but should not be able to publish on their own) but some are also Reviewers (can edit AND publish). The point of this would be to allow full editability in the sandbox environment, but have a quality gate before it's shared with the wider world.
Almost all content should go through this same quality gate, ideally except mostly for corrections of typos or any other really time-critical edits. A much more straightforward approach is that literally ALL edits must be accepted by a Reviewer before they can go live.

**Comment by scottgallant** [comment]
I just played around with Vercel's new [commenting features](https://vercel.com/blog/introducing-commenting-on-preview-deployments) for preview deployments. It's pretty impressive. 

Will our suggested 2-branch (staging and production) workflow make it awkward for our users to use Vercel Commenting with Tina?   Is there a way we can make this a tighter, more intuitive integration?

**Comment by dasmeet** [comment]
There is this workaround I am planning to implement on my site. In case, it helps.

I have a 11ty blog setup with Tina.

My plan is:
Disable auto-build on Git push. 
Use Tina to make and save changes as usual. 
When I am done, I manually trigger build using webhooks. 

If there was a Publish button in Tina that could trigger a POST webhook, even better.

**Comment by cascading-jox** [comment]
Are there any updates on this? This is a feature my client would greatly appreciate 😊

**Comment by joacimeldre** [comment]
I would also like the ability to save multiple new, edited and deleted content before publishing. Is this something you're looking into developing into a smoother workflow than Git branches for TinaCMS?
